### PR TITLE
devMode Integration

### PR DESCRIPTION
**Pull request by lpsandaruwan**
* Based on the implementation of this [Colyseus.js change](https://github.com/colyseus/colyseus.js/commit/43d5ce28bdb9676eeb450e0638d7480f197e39b1).
* Test materials have been added to this repository, https://github.com/lpsandaruwan/colyseus-devmode-for-unity-sample

### CODE REVIEW

File: Assets/Colyseus/Runtime/Scripts/Models/ColyseusClient.cs [hunk 1]
```diff
@@ -284,11 +285,17 @@ public async Task<T[]> GetAvailableRooms<T>(string roomName = "", Dictionary<str
         /// </summary>
         /// <param name="response">The response from the matchmaking attempt</param>
         /// <param name="headers">Dictionary of headers to pass to the server</param>
+        /// <param name="devModeRetryAttempt">Index to keep devMode re-connection tryout number: Please do not use this devMode param for general purposes</param>
+        /// <param name="previousRoom">Previous ColyseusRoom{T} instance to re-establish the server connection: Please do not use this devMode param for general purposes</param>
         /// <typeparam name="T">Type of <see cref="ColyseusRoom{T}" /> we're consuming the seat from</typeparam>
         /// <returns><see cref="ColyseusRoom{T}" /> in which we now have a seat via async task</returns>
         public async Task<ColyseusRoom<T>> ConsumeSeatReservation<T>(ColyseusMatchMakeResponse response,
-            Dictionary<string, string> headers = null)
+            Dictionary<string, string> headers = null, int devModeRetryAttempt = 0, ColyseusRoom<T> previousRoom = null)
```
Comment (endel) [inline Assets/Colyseus/Runtime/Scripts/Models/ColyseusClient.cs:293]: remove `devModeRetryAttempt` parameter

File: Assets/Colyseus/Runtime/Scripts/Models/ColyseusClient.cs [hunk 2]
```diff
@@ -304,29 +311,57 @@ public async Task<T[]> GetAvailableRooms<T>(string roomName = "", Dictionary<str
                 queryString.Add("reconnectionToken", room.ReconnectionToken);
             }
 
-            room.SetConnection(CreateConnection(response.room, queryString, headers));
+            ColyseusRoom<T> targetRoom = previousRoom ?? room;
+            if (response.devMode) // Integrate devMode server re-connection callback
+            {
+	            room.SetConnection(CreateConnection(response.room, queryString, headers), async () =>
+	            {
+		            LSLog.Log($"[Colyseus devMode]: Re-establishing connection with room id {room.RoomId}");
```
Comment (endel) [inline Assets/Colyseus/Runtime/Scripts/Models/ColyseusClient.cs:319]: need to keep track of the number of retries within this scope - i suggest creating a local `Action` here, similarly to how this is approached in the javascript implementation.

**Review by endel — COMMENTED** [review]
Hi @lpsandaruwan, the reconnection is working as expected, but I noticed a problem with the retry logic. 

The retry counter should be re-set after the connection has been successfully reestablished. 

In the following screenshot, you can see that the retry count builds up even after successful reconnects:

image

**Comment by lpsandaruwan** [comment]
> Hi @lpsandaruwan, the reconnection is working as expected, but I noticed a problem with the retry logic.
> 
> The retry counter should be re-set after the connection has been successfully reestablished.
> 
> In the following screenshot, you can see that the retry count builds up even after successful reconnects:
> 
> image

Hi @endel,

Did a workaround for fixing this. But Now the logic has a infinite depth to recursion. I'm working on it.

File: Assets/Colyseus/Runtime/Scripts/Models/ColyseusClient.cs [hunk 3]
```diff
@@ -304,29 +306,66 @@ public async Task<T[]> GetAvailableRooms<T>(string roomName = "", Dictionary<str
...
+			            }
+			            else
+			            {
+				            LSLog.Log(
+					            $"[Colyseus devMode]: Failed to reconnect! Is your server running? Please check server logs!");
+			            }
+		            }
+	            }
+
+	            Thread.Sleep(1000);
```
Comment (endel) [inline Assets/Colyseus/Runtime/Scripts/Models/ColyseusClient.cs:342]: https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.delay?view=net-6.0

**Comment by endel** [comment]
This is great! Thanks @lpsandaruwan 🎉

**Pull request merged**